dbms/src/Debug/astToExecutor.cpp

```diff
@@ -1648,30 +1657,64 @@ ExecutorPtr compileProject(ExecutorPtr input, size_t & executor_index, ASTPtr se
     return project;
 }
 
-ExecutorPtr compileJoin(size_t & executor_index, ExecutorPtr left, ExecutorPtr right, ASTPtr params)
+ExecutorPtr compileJoin(size_t & executor_index, ExecutorPtr left, ExecutorPtr right, tipb::JoinType tp, ASTPtr using_expr_list)
 {
     DAGSchema output_schema;
-    const auto & join_params = (static_cast<const ASTTableJoin &>(*params));
+
     for (auto & field : left->output_schema)
     {
-        if (join_params.kind == ASTTableJoin::Kind::Right && field.second.hasNotNullFlag())
+        if (tp == tipb::JoinType::TypeRightOuterJoin && field.second.hasNotNullFlag())
             output_schema.push_back(toNullableDAGColumnInfo(field));
         else
             output_schema.push_back(field);
     }
-    for (auto & field : right->output_schema)
+
+    /// for any kind of semi join, the right table column is ignored
+    /// besides, for anti left outer semi join and left outer semi join a boolean field is pushed back
+    if (tp == tipb::JoinType::TypeLeftOuterSemiJoin || tp == tipb::JoinType::TypeAntiLeftOuterSemiJoin)
     {
-        if (join_params.kind == ASTTableJoin::Kind::Left && field.second.hasNotNullFlag())
-            output_schema.push_back(toNullableDAGColumnInfo(field));
-        else
-            output_schema.push_back(field);
+        // todo: figure out the correct way to set column info
+        auto ci = ColumnInfo{};
+        output_schema.push_back(std::make_pair("", ci));
```

Comment:
why are these codes?

Reply:
It seems for leftSemiFamily join an additional column will be used

https://github.com/pingcap/tiflash/blob/72f7481fe641f3c3dc43bdc5c4984ac23fb8f45b/dbms/src/Flash/Coprocessor/JoinInterpreterHelper.cpp#L185-L199

https://github.com/pingcap/tiflash/blob/72f7481fe641f3c3dc43bdc5c4984ac23fb8f45b/dbms/src/Flash/Coprocessor/JoinInterpreterHelper.cpp#L306-L309

Also according to the comment, the additional column is used to indicate whether a row matches.

https://github.com/pingcap/tiflash/blob/72f7481fe641f3c3dc43bdc5c4984ac23fb8f45b/dbms/src/Parsers/ASTTablesInSelectQuery.h#L104-L105

Reply:
IMO, we can ignore match_helper_column here.
Because this column will not be referenced by the upper operator, nor will it be generated by the lower operator.
It's just a dross in the join implementation~

Reply:
Sorry, I got it wrong, it should be a match_helper_column...

Reply:
I think so.
Like filter_column_name in FilterBlockInputStream

Reply:
It's Block/Stream's schema not Operator's schema

Reply:
sure

Reply:
reached a consensus: this column will be used by upper operator.